### PR TITLE
05_creating_CNN

### PR DESCRIPTION
**Pull request by Dimonovez**
One universal notebook without learning rate decay.

### CODE REVIEW

**Comment by roman-kh** [comment]
Global variables are a malpractice. Use pipeline configs instead

**Comment by roman-kh** [comment]
Such simple `ConvModel` might also be omitted in favour of model config

**Comment by Dimonovez** [comment]
We want to use one universal notebook for tf and torch? @roman-kh

File: examples/tutorials/05_creating_CNN.ipynb [hunk 1]
```diff
@@ -13,8 +13,7 @@
    "source": [
```
Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Use shorter names: model, channels</p><p>image_shape is not needed at all</p><p> </p><p>Moreover, you don't need to choose framework either. Just state the model used.</p><p>Write a detailed description about models and frameworks, provide a short inheritance example</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>B.image_shape</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>before pipeline is not needed here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>before pipeline is not needed here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>before pipeline is not needed here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Use util graph function</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

File: examples/utils/utils.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import matplotlib.pyplot as plt
 import seaborn as sns
 import numpy as np
-from pandas import ewma
+import pandas as pd
```
Comment (SergeyTsimfer) [inline examples/utils/utils.py]: Why?

Comment (Dimonovez) [reply examples/utils/utils.py]: Error in draw function because of the removal of ewma.

File: examples/tutorials/05_creating_CNN.ipynb (hunk 1 above)
Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Extremely unclear explanation!!!</p><p>Why anybody need to open the source of base TFModel or TorchModel?</p><p>You mention models, then TF and Torch, then again models and again Torch and TF, then models and once more TF and Torch. That's extremely confusing.</p><p>Write clearly, concisely and cohesively about models, TF / Torch difference and framework-agnosticism.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>There are no random_rotate and random_scale</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Comment by SergeyTsimfer** [comment]
[universal 05 tutorial](https://trello.com/c/bBt8yAFS/30-universal-05-tutorial)

File: examples/tutorials/05_creating_CNN.ipynb (hunk 1 above)
Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Maybe, append to <code>sys.path</code>  once?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>The next cell has nothing to do with batch classes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Make the first passage more fluid and less declarative. Currently it states incorrect information about lots of things: for example, TensorFlow can make dynamic graph.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>The next cell has nothing to do with <code>default_config</code>,  and you don't really need to know about its existence.</p><p>Describe what <code>layout</code>  is, and what different letters stand for</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>It is a function from current namespace, is not it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Do we really want test time augmentations?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (Dimonovez) [reply examples/tutorials/05_creating_CNN.ipynb]: <p>Yes!</p>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>You don't need to implement a MNIST dataset. It is already done for you</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>The last line looks clunky</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p><code>And 'body/strides' is an integer -- therefore, the same value is used for all convolutional layers</code></p><p> </p><p><code>Explain more keys: at least all of the used in the next cell: output, </code> initial_block/inputs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (SergeyTsimfer) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>split long lines</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (Dimonovez) [reply examples/tutorials/05_creating_CNN.ipynb]: Too much information for this tutorial.

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>from batchflow.utils import plot_images</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>But before, we should define model class&nbsp;<code>'model'</code>&nbsp;and channels positions&nbsp;<code>'channels'</code>&nbsp;(for TensorFlow models -&nbsp;<strong>'last'</strong>, for PyTorch models&nbsp;- <strong>'first'</strong>) in&nbsp;<code>config</code>.</p><p> </p><p>There are also predefined&nbsp;<a href="https://analysiscenter.github.io/batchflow/intro/model_zoo.html" target="_blank">models</a>&nbsp;of both frameworks. You can use them without additional configuration.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Give an example for tf models</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Do we need plt.rc?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

**Comment by roman-kh** [comment]
And add some headers for a better structure

File: examples/tutorials/05_creating_CNN.ipynb [hunk 3]
```diff
@@ -12,9 +12,8 @@
    "metadata": {},
```
Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Remove targets from predict_model</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p><code>window</code> is a constant. give it a proper name and use CAPITALS</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Mention that there is no augmentation this time</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Rename to inference_pipeline</p><p>Remove targets from predict_model</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (roman-kh) [inline examples/tutorials/05_creating_CNN.ipynb]: <p>Why is it not shift and flattened any more?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/analysiscenter/batchflow/pull/476/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (Dimonovez) [reply examples/tutorials/05_creating_CNN.ipynb]: Cause it was a data augmentation method. This the inference pipeline without such methods. 
Also, now there is no `shift_flattened` function. This function shifted an image as well as the shift method of ImagesBatch. Now in the notebook used shift method of ImagesBatch (replace of shift_flattened for image shifting) and custom_filter (replace of shift_flattened as a user-defined function for data augmentation).

Comment (Dimonovez) [reply examples/tutorials/05_creating_CNN.ipynb]: Now, there are two parameters FREQUENCY and N_LAST and they are equal. 
FREQUENCY is responsible for how often evaluate metrics.
N_LAST  is responsible for a selected number of last results for evaluation.
